### DISCUSSION TITLE
Iterative Pricing Exercise

### DISCUSSION THREAD

**Discussion by mhumpher** [fumitoh/modelx#66] [Q&A] [answered]
I load charges/premiums as new cells into one of my spaces. However, as part of pricing exercises we might want to iteratively test different sets of charges/premiums. Is there a way to replace or update input cells?

**Comment by fumitoh** [comment, marked as answer]
You can assign input values to cells by ``[] =``

For example,

```py
import modelx as mx
from modelx import defcells

@defcells
def pv_premiums():
    return 0 # dummy

@defcells
def pv_claims():
    return 3

@defcells
def net_pv():
    return pv_premiums() - pv_claims()
``` 

and

```py

>>> pv_premiums[()] = 4

>>> net_pv()
1

>>> pv_premiums[()] = 3

>>> net_pv()
0
```

You can write a loop in Python to iterate the assignment operation above.  but it's more recommended to refer to a Reference instead of directly assigning values to `pv_premiums`:

```py
@defcells
def pv_premiums():
    return pv_prem_input
```

```py
>>> mx.Model1.Space1.pv_prem_input = 4

>>> net_pv()
1

>>> mx.Model1.Space1.pv_prem_input = 3

>>> net_pv()
0
```

**Reply by fumitoh** [reply]
Or, you can even write a solver using `minimize_scalar` from `scipy`.

 ```py
from scipy.optimize import minimize_scalar  

def solv_premium(space):
    
    def solv_prem_inner(x):
        space.pv_prem_input  = x
        return space.net_pv()**2
    
    res = minimize_scalar(solv_prem_inner)
    return res.x
```

```py
>>> solv_premium(mx.Model1.Space1)
3.0

>>> mx.Model1.Space1.pv_prem_input
2.99999995559
```

Furture more, you can write the solver code above in a *.py file and save it within the model using `new_module`.
